.doom.d/config.org
#+TITLE: Doom Emacs Config
#+PROPERTY: header-args :tangle yes

* Table of contets :toc:
- [[#about-this-config][About this config]]
- [[#dired][dired]]
- [[#doom-theme][Doom theme]]
- [[#fonts][Fonts]]
- [[#line-settings][Line settings]]
- [[#mu4e][MU4E]]
  - [[#falta-configurar-esse][Falta configurar esse]]
- [[#neotree][Neotree]]
- [[#open-specific-files][Open specific files]]
- [[#shells][Shells]]
- [[#org-mode][Org mode]]
  - [[#getting-started][Getting started]]
  - [[#tasks-and-states][Tasks and States]]
  - [[#adding-new-tasks-quickly-with-org-capture][Adding New Tasks Quickly with Org Capture]]
  - [[#refiling-tasks][Refiling Tasks]]
  - [[#custom-agenda-views][Custom Agenda Views]]
  - [[#configs-que-ja-tinha-feito-antes-de-ler-o-guia][Configs que ja tinha feito antes de ler o guia]]

* About this config
This is my personal Doom Emacs config.

To use this file you can load in Emacs an tangle with =C-c C-v C-t= which will
create config.el in the same directory as this file. This will extract all elisp
blocks in this document into a file to be used as Emacs config file.

This document also describes my workflow.

Based on:
- [[https://gitlab.com/dwt1/dotfiles/-/blob/master/.doom.d/config.org]]
- [[http://doc.norang.ca/org-mode.html][Org Mode - Organize Your Life In Plain Text!]]

* dired
Dired is the file manager within Emacs. Below, I setup keybindings for image
previews (peep-dired). Doom Emacs does not use 'SPC d' for any of its
keybindings, so I've chosen the format of 'SPC d' plus 'key'.

| COMMAND                                   | DESCRIPTION                                     | KEYBINDING |
|-------------------------------------------+-------------------------------------------------+------------|
| dired                                     | /Open dired file manager/                         | SPC d d    |
| dired-jump                                | /Jump to current directory in dired/              | SPC d j    |
| (in dired) peep-dired                     | /Toggle image previews within dired/              | SPC d p    |
| (in dired) dired-view-file                | /View file in dired/                              | SPC d v    |
| (in dired) dired-up-directory             | Go up in the directory tree                     | h          |
| (in dired) dired-find-file                | Go down in the directory tree (or open if file) | l          |
| (in peep-dired-mode) peep-dired-next-file | /Move to next file in peep-dired-mode/            | j          |
| (in peep-dired-mode) peep-dired-prev-file | /Move to previous file in peep-dired-mode/        | k          |

#+BEGIN_SRC emacs-lisp
(map! :leader
      :desc "Dired"
      "d d" #'dired
      :leader
      :desc "Dired jump to current"
      "d j" #'dired-jump
      (:after dired
        (:map dired-mode-map
         :leader
         :desc "Peep-dired image previews"
         "d p" #'peep-dired
         :leader
         :desc "Dired view file"
         "d v" #'dired-view-file)))
;; Make 'h' and 'l' go back and forward in dired. Much faster to navigate the directory structure!
(evil-define-key 'normal dired-mode-map
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file) ; use dired-find-file instead if not using dired-open package
;; If peep-dired is enabled, you will get image previews as you go up/down with 'j' and 'k'
(evil-define-key 'normal peep-dired-mode-map
  (kbd "j") 'peep-dired-next-file
  (kbd "k") 'peep-dired-prev-file)
(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
;; Get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
;; With dired-open plugin, you can launch external programs for certain extensions
;; For example, I set all .png files to open in 'sxiv' and all .mp4 files to open in 'mpv'
(setq dired-open-extensions '(("gif" . "sxiv")
                              ("jpg" . "sxiv")
                              ("png" . "sxiv")
                              ("mkv" . "mpv")
                              ("mp4" . "mpv")))
#+END_SRC

* Doom theme
Set doom theme.

#+begin_src emacs-lisp
(setq doom-theme 'doom-dracula)
#+end_src

* Fonts
Settings related to fonts within Doom Emacs:
- 'doom-font' -- standard monospace font that is used for most things in Emacs.
- 'doom-variable-pitch-font' -- variable font which is useful in some Emacs plugins.
- 'doom-big-font' -- used in doom-big-font-mode; useful for presentations.
- 'font-lock-comment-face' -- for comments.
- 'font-lock-keyword-face' -- for keywords with special significance, like ‘for’ and ‘if’ in C.

#+begin_src emacs-lisp
;; (setq doom-font (font-spec :family "DejaVu Sans Mono" :size 16)
(setq doom-font (font-spec :family "SauceCodePro Nerd Font Mono" :size 16 :style 'Medium)
      doom-variable-pitch-font (font-spec :family "Ubuntu" :size 15)
      doom-big-font (font-spec :family "DejaVu Sans Mono" :size 24))

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

;; Show comments in italic
(custom-set-faces!
  '(font-lock-comment-face :slant italic))
#+end_src

* Line settings
Toggle display-line-numbers-type so I have line numbers relative to the current
line. Doom Emacs uses 'SPC t' for "toggle" commands, so I choose 'SPC t t' for
toggle-truncate-lines.
#+begin_src emacs-lisp
(setq display-line-numbers-type 'relative)
(map! :leader
      :desc "Truncate lines"
      "t t" #'toggle-truncate-lines)
#+end_src

* MU4E
Setting up mu4e which is an email client that works within emacs. You must
install mu4e and mbsync through your Linux distribution's package manager.
Setting up smtp for sending mail. Make sure the gnutls command line utils are
installed. Package 'gnutls-bin' in Debian/Ubuntu, and 'gnutls' in Arch.

** TODO Falta configurar esse

* Neotree
Neotree is a file tree viewer. When you open neotree, it jumps to the current
file thanks to neo-smart-open. The neo-window-fixed-size setting makes the
neotree width be adjustable. Doom Emacs had no keybindings set for neotree.
Since Doom Emacs uses 'SPC t' for 'toggle' keybindings, I used 'SPC t n' for
toggle-neotree.

#+begin_src emacs-lisp
(after! neotree
  (setq neo-window-fixed-size nil))
#+end_src

* Open specific files
Keybindings to open files that I work with all the time using the find-file
command non-interactively since we specify exactly what file to open. The format
I use for these bindings is 'SPC -' plus 'key' since Doom Emacs does not use
these keybindings.

| PATH TO FILE     | DESCRIPTION      | KEYBINDING |
|------------------+------------------+------------|
| ~/org/agenda.org | /Edit agenda file/ | SPC - a    |

#+begin_src emacs-lisp
(map! :leader
      :desc "Edit agenda file"
      "a" #'(lambda () (interactive) (find-file "~/org/todo.org")))
#+end_src

* Shells
Settings for the various shells and terminal emulators within Emacs.
- 'shell-file-name' -- sets the shell to be used in M-x shell, M-x term, M-x
  ansi-term and M-x vterm.
- 'eshell-aliases-file' -- sets an aliases file for the eshell.

#+BEGIN_SRC emacs-lisp
(setq shell-file-name "/bin/bash"
      eshell-aliases-file "~/.doom.d/aliases"
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "htop" "ssh" "zsh")
      vterm-max-scrollback 5000)
#+end_src

* Org mode
** Getting started

*** Organizing My Life into Org Files
Tasks are separated into logical groupings or projects. Use separate org files
for large task groupings and subdirectories for collections of files for
multiple projects that belong together.

Here are the sample files that I use
| Filename                  | Description                                |
|---------------------------+--------------------------------------------|
| todo.org                  | Personal tasks and things to keep track of |
| inbox.org                 | Capture inbox                              |
| aurearobotics/Foo/bar.org | Taks for project Bar of client Foo         |

Org-mode is great for dealing with multiple clients and client projects. An org
file becomes the collection of notes, tasks, reference material, etc. for a
single client-project.

Separating the tasks for each client-project into separate org files helps keep
things logically grouped and since projects come and go, this allows entire org
files to be added or dropped from my agenda to keep only what is important
visible in agenda views.

*** Agenda
This is the =org-agenda-files= setup.

#+begin_src emacs-lisp
(after! org
  (setq org-agenda-files '("~/org/"
                           "~/org/aurearobotics"
                           "~/org/aurearobotics/firjan"
                           "~/org/aurearobotics/brunel")))
#+end_src

=org-mode= manages the =org-agenda-files= variable automatically using =C-c [= and =C-c
]= to add and remove files respectively. However, this replaces my directory list
with a list of explicit filenames instead and is not what I want. If this occurs
then adding a new org file to any of the above directories will not contribute
to my agenda and I will probably miss something important.

I have disabled the =C-c [= and =C-c ]= keys in =org-mode-hook= to prevent messing up
my list of directories in the =org-agenda-files= variable. I just add and remove
directories manually above. Changing the list of directories in =org-agenda-files=
happens very rarely since new files in existing directories are automatically
picked up.

#+begin_src emacs-lisp
(after! org
    (define-key org-mode-map (kbd "C-c [") nil)
    (define-key org-mode-map (kbd "C-c ]") nil))
#+end_src

*** Org File Structure
Most of my org files are set up with level 1 headings as main categories only.
Tasks and projects normally start as level 2.

Here are some examples of my level 1 headings in

=todo.org=:
- Special Dates
  + Birthdays
  + Holidays
- Finances
- Health
- House Maintenance
- Notes
- Tasks

=aurearobotics.org=
- System Maintenance
- Payroll
- Accounting
- Finances
- Hardware Maintenance
- Tasks
- Research and Development
- Notes

*** Key bindings
The main reason I have special key bindings (like =F11=, and =F12=) is so that the
keys work in any mode. If I'm in the Gnus summary buffer then =C-u C-c C-x C-i=
doesn't work, but the =C-F11= key combination does and this saves me time since I
don't have to visit an org-mode buffer first just to clock in a recent task.

** Tasks and States

I use the same set of TODO keywords for all of my org files. Org-mode lets you
define TODO keywords per file but I find it's easier to have a standard set of
TODO keywords globally so I can use the same setup in any org file I'm working
with.

*** Todo keywords
#+begin_src emacs-lisp
(after! org
  (setq org-todo-keywords
        '((sequence "TODO(t)" "NEXT(n)" "WAITING(w@/!)" "|" "DONE(d)" "CANCELLED(c@/!)" "PHONE" "MEETING"))))
        ;; Antes de mudar as cores, dar um describe nesta variavel e só entao adaptar a partir dela
        ;; org-todo-keyword-faces
        ;; '(("TODO" :foreground "red" :weight bold)
        ;;    ("NEXT" :foreground "blue" :weight bold)
        ;;    ("DONE" :foreground "forest green" :weight bold)
        ;;    ("WAITING" :foreground "orange" :weight bold)
        ;;    ("HOLD" :foreground "magenta" :weight bold)
        ;;    ("CANCELLED" :foreground "forest green" :weight bold)
        ;;    ("MEETING" :foreground "forest green" :weight bold)
        ;;    ("PHONE" :foreground "forest green" :weight bold))))
#+end_src

*** Project Task States
I don't like to bother with manually stating 'this is a project' and 'that is
not a project'. For me a project definition is really simple. If a task has
subtasks with a todo keyword then it's a project. That's it.

One subtask of a project needs to be marked NEXT so the project is not on the
stuck projects list.

*** Phone Calls
Telephone calls are special. They are created in a done state by a capture task.

The time of the call is recorded for as long as the capture task is active. If I
need to look up other details and want to close the capture task early I can
just =C-c C-c= to close the capture task (stopping the clock) and then =f9 SPC= to
resume the clock in the phone call while I do other things.

*** Meetings
Meetings are special. They are created in a done state by a capture task. I use
the MEETING capture template when someone interrupts what I'm doing with a
question or discussion. This is handled similarly to phone calls where I clock
the amount of time spent with whomever it is and record some notes of what was
discussed, either during or after the meeting.

The time of the meeting is recorded for as long as the capture task is active.
If I need to look up other details and want to close the capture task early I
can just =C-c C-c= to close the capture task (stopping the clock) and then =f9 SPC=
to resume the clock in the meeting task while I do other things.

*** Fast Todo Selection
Fast todo selection allows changing from any task todo state to any other state
directly by selecting the appropriate key from the fast todo selection key menu.
This is a great feature!

#+begin_src emacs-lisp
(setq org-use-fast-todo-selection t)
#+end_src

Changing a task state is done with =SPC m t KEY= where =KEY= is the appropriate fast
todo state selection key as defined in =org-todo-keywords=.

The setting

#+begin_src emacs-lisp
(setq org-treat-S-cursor-todo-selection-as-state-change nil)
#+end_src

allows changing todo states with =S-left= and =S-right= skipping all of the normal
processing when entering or leaving a todo state. This cycles through the todo
states but skips setting timestamps and entering notes which is very convenient
when all you want to do is fix up the status of an entry.

** Adding New Tasks Quickly with Org Capture
I create new tasks quickly using these templates. They are store in =inbox.org=
and them I refile them to the correct location. Normally this file is empty
because I refile my inbox frequently.

*** Capture Templates

When a new item needs to be added I categorize it into one of a few things:

- A phone call (=p)=
- A meeting (=m=)
- An email I need to respond to (=r=)
- A new task (=t=)
- A new note (=n=)
- A journal entry (=j=)
- A new habit (=h=)

Here is my setup for org-capture

#+begin_src emacs-lisp
(after! org
  (setq org-directory "~/org"
        org-default-notes-file "~/org/inbox.org"
        +org-capture-todo-file "~/org/inbox.org"
        +org-capture-journal-file "~/org/journal.org"))

;; I use C-c c to start capture mode
(global-set-key (kbd "C-c c") 'org-capture)

(after! org
  (setq org-capture-templates
        '(("t" "Todo" entry (file +org-capture-todo-file)
           "* TODO %?\n:PROPERTIES:\n:CREATED: %U\n:END:\n")
          ("r" "Responder Email" entry (file +org-capture-todo-file)
               "* NEXT Respond %:from on %:subject\nSCHEDULED: %t\n:PROPERTIES:\n:CREATED: %U\n:END:\n%a\n" :immediate-finish t)
          ("n" "Note" entry (file org-default-notes-file)
               "* %? :NOTE:\n:PROPERTIES:\n:CREATED: %U\n:END:\n%a\n" :clock-in t :clock-resume t)
          ("j" "Journal" entry (file+olp-datetree +org-capture-journal-file)
               "* %U %?\n" :clock-in t :clock-resume t)
          ("w" "org-protocol" entry (file +org-capture-todo-file)
               "* TODO Review %c\n:CREATED: %U\n:END:\n" :immediate-finish t)
          ("m" "Meeting" entry (file +org-capture-notes-file)
               "* MEETING with %?\n:CREATED: %U\n:END:\n" :clock-in t :clock-resume t)
          ("p" "Phone call" entry (file +org-capture-notes-file)
               "* PHONE call with %?\n:CREATED: %U\n:END:\n" :clock-in t :clock-resume t))))
#+end_src

Capture mode handles automatically clocking in and out of a capture task. When I
start a capture mode task the task is clocked in as specified by =:clock-in t= and
when the task is filed with =C-c C-c= the clock resumes on the original clocking
task.

I use journal entries to log interruptions as events for that day.

** Refiling Tasks
Refiling tasks is easy. After collecting a bunch of new tasks in my refile.org
file using capture mode I need to move these to the correct org file and topic.
All of my active org-files are in my =org-agenda-files= variable and contribute to
the agenda.

I collect capture tasks in inbox.org for up to a week. These stand out daily on
my block agenda and I usually refile them during the day. I like to keep my
refile task list empty.

*** Refile Setup
To refile tasks in org you need to tell it where you want to refile things.

In my setup I let any file in =org-agenda-files= and the current file contribute
to the list of valid refile targets.

#+begin_src emacs-lisp
; Targets include this file and any file contributing to the agenda - up to 9 levels deep
(after! org
  (setq org-refile-targets '((nil :maxlevel . 9) (org-agenda-files :maxlevel . 9))
        ; Use full outline paths for refile targets - we file directly with IDO
        org-refile-use-outline-path t
        ; Targets complete directly with IDO
        org-outline-path-complete-in-steps nil
        ; Allow refile to create parent tasks with confirmation
        org-refile-allow-creating-parent-nodes (quote confirm)))

; Exclude DONE state tasks from refile targets
(defun sr/verify-refile-target ()
  "Exclude todo keywords with a done state from refile targets"
  (not (member (nth 2 (org-heading-components)) org-done-keywords)))

(after! org
  (setq org-refile-target-verify-function 'sr/verify-refile-target))
#+end_src

To refile a task to my =inbox.org= file under =System Maintenance= I just put the
cursor on the task and hit =C-c C-w= and enter =main RET= and it's done.

*** Refiling Tasks
Tasks to refile are in their own section of the block agenda. To find tasks to
refile I run my agenda view with =F12 SPC= and scroll down to second section of
the block agenda: =Tasks to Refile=. This view shows all tasks (even ones marked
in a =done= state).

Bulk refiling in the agenda works very well for multiple tasks going to the same
place. Just mark the tasks with =m= and then =B r= to refile all of them to a new
location. Occasionally I'll also refile tasks as subtasks of the current
clocking task using =C-2 C-c C-w= from the =inbox.org= file.

Refiling all of my tasks tends to take less than a minute so I normally do this
a couple of times a day.

*** Refiling Notes
I keep a =* Notes= headline in most of my org-mode files. Notes have a =NOTE= tag
which is created by the capture template for notes. This allows finding notes
across multiple files easily using the agenda search functions.

Notes created by capture tasks go first to =inbox.org= and are later refiled to
the appropriate project file. Some notes that are project related get filed to
the appropriate project instead of under the catchall =* NOTES= task. Generally
these types of notes are specific to the project and not generally useful – so
removing them from the notes list when the project is archived makes sense.

*** Refiling Phone Calls and Meetings
Phone calls and meetings are handled using capture mode. I time my calls and
meetings using the capture mode template settings to clock in and out the
capture task while the phone call or meeting is in progress.

Phone call and meeting tasks are collected in =inbox.org= and are later refiled to
the appropriate location. I refile my phone call and meeting tasks under the
appropriate project so time tracking and reports are as accurate as possible.

** Custom Agenda Views
I have one block agenda view that has everything on it.

Custom agenda views are used for:
- Single block agenda shows the following
  + Overview of today
  + Finding tasks to be refiled
  + Finding stuck projects
  + Finding NEXT tasks to work on
  + Show all related tasks
  + Reviewing projects
  + Finding tasks waiting on something
  + Finding tasks to be archived
- Finding notes

If I want just today's calendar view then =F12= a is still faster than generating
the block agenda - especially if I want to view a week or month's worth of
information, or check my clocking data. In that case the extra detail on the
block agenda view is never really needed and I don't want to spend time waiting
for it to be generated.

*** Setup
#+begin_src emacs-lisp
;; Do not dim blocked tasks
(setq org-agenda-dim-blocked-tasks nil)

;; Compact the block agenda view
(setq org-agenda-compact-blocks t)

;; Custom agenda command definitions
(setq org-agenda-custom-commands
      '(("N" "Notes" tags "NOTE"
         ((org-agenda-overriding-header "Notes")
          (org-tags-match-list-sublevels t)))
        ("h" "Habits" tags-todo "STYLE=\"habit\""
         ((org-agenda-overriding-header "Habits")
          (org-agenda-sorting-strategy
           '(todo-state-down effort-up category-keep))))
        (" " "Agenda"
         ((agenda "" nil)
          (tags "REFILE"
                ((org-agenda-overriding-header "Tasks to Refile")
                 (org-tags-match-list-sublevels nil))))
         nil)))
#+end_src

This shows top-level projects and =NEXT= tasks but hides the project details since
we are not focused on any particular project.

I generally work top-down on the agenda. Things with deadlines and scheduled
dates show up in the agenda at the top.

My day goes generally like this:

- Punch in
- Look at the agenda and make mental note of anything important to deal with today
- Start working on a task
  + Clock it in
  + Work on it until it is =DONE= or it gets interrupted
- Make journal entries (=C-c c j=) for interruptions
- Punch out for breaks and punch back after breaks
- Refile tasks to empty the list
  + Tag tasks to be refiled with =m= collecting all tasks for the same target
  + Bulk refile the tasks on the target location with =B r=
  + Refile individually with =C-c C-w=
- Punch out

**** TODO Configurar =org-agenda-custom-commands= depois de já ter criado e entendido como funciona e a necessidade dos métodos que ele cria

*** Reading email, articles and conversations
When reading email I just let the default task clock the time I spend on these
tasks. To read email I go to the Emacs email client and read everything in my
inboxes. If there are emails that require a response I use org-capture to create
a new task with a heading of 'Respond to <user>' for each one. This
automatically links to the email in the task and makes it easy to find later.
Some emails are quick to respond to and some take research and a significant
amount of time to complete. I clock each one in it's own task. The capture
template for Respond To tasks is schedule for today so I can refile the task to
the appropriate org file without losing the task and end up forgeting about it.

Next, I go to my newly created tasks to be refiled from the block agenda and
clock in an email task and deal with it.

If an article has a useful piece of information I want to remember I create a
note for it with =C-c c n= and enter the topic and file it. This takes practically
no time at all and I know the note is safely filed for later retrieval. The time
I spend in the capture buffer is clocked with that capture note.

** Configs que ja tinha feito antes de ler o guia
*** Outras configs do org
#+begin_src emacs-lisp
(after! org
  (require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-ellipsis " ▼ "
        org-log-done 'time
        org-hide-emphasis-markers t
        calendar-week-start-day 1
        ))
#+end_src


.doom.d/config.el
(map! :leader
      :desc "Dired"
      "d d" #'dired
      :leader
      :desc "Dired jump to current"
      "d j" #'dired-jump
      (:after dired
        (:map dired-mode-map
         :leader
         :desc "Peep-dired image previews"
         "d p" #'peep-dired
         :leader
         :desc "Dired view file"
         "d v" #'dired-view-file)))
;; Make 'h' and 'l' go back and forward in dired. Much faster to navigate the directory structure!
(evil-define-key 'normal dired-mode-map
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file) ; use dired-find-file instead if not using dired-open package
;; If peep-dired is enabled, you will get image previews as you go up/down with 'j' and 'k'
(evil-define-key 'normal peep-dired-mode-map
  (kbd "j") 'peep-dired-next-file
  (kbd "k") 'peep-dired-prev-file)
(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
;; Get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
;; With dired-open plugin, you can launch external programs for certain extensions
;; For example, I set all .png files to open in 'sxiv' and all .mp4 files to open in 'mpv'
(setq dired-open-extensions '(("gif" . "sxiv")
                              ("jpg" . "sxiv")
                              ("png" . "sxiv")
                              ("mkv" . "mpv")
                              ("mp4" . "mpv")))

(setq doom-theme 'doom-dracula)

;; (setq doom-font (font-spec :family "DejaVu Sans Mono" :size 16)
(setq doom-font (font-spec :family "SauceCodePro Nerd Font Mono" :size 16 :style 'Medium)
      doom-variable-pitch-font (font-spec :family "Ubuntu" :size 15)
      doom-big-font (font-spec :family "DejaVu Sans Mono" :size 24))

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

;; Show comments in italic
(custom-set-faces!
  '(font-lock-comment-face :slant italic))

(setq display-line-numbers-type 'relative)
(map! :leader
      :desc "Truncate lines"
      "t t" #'toggle-truncate-lines)

(after! neotree
  (setq neo-window-fixed-size nil))

(map! :leader
      :desc "Edit agenda file"
      "a" #'(lambda () (interactive) (find-file "~/org/todo.org")))

(setq shell-file-name "/bin/bash"
      eshell-aliases-file "~/.doom.d/aliases"
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "htop" "ssh" "zsh")
      vterm-max-scrollback 5000)

(after! org
  (setq org-agenda-files '("~/org/"
                           "~/org/aurearobotics"
                           "~/org/aurearobotics/firjan"
                           "~/org/aurearobotics/brunel")))

(after! org
    (define-key org-mode-map (kbd "C-c [") nil)
    (define-key org-mode-map (kbd "C-c ]") nil))

(after! org
  (setq org-todo-keywords
        '((sequence "TODO(t)" "NEXT(n)" "WAITING(w@/!)" "|" "DONE(d)" "CANCELLED(c@/!)" "PHONE" "MEETING"))))
        ;; Antes de mudar as cores, dar um describe nesta variavel e só entao adaptar a partir dela
        ;; org-todo-keyword-faces
        ;; '(("TODO" :foreground "red" :weight bold)
        ;;    ("NEXT" :foreground "blue" :weight bold)
        ;;    ("DONE" :foreground "forest green" :weight bold)
        ;;    ("WAITING" :foreground "orange" :weight bold)
        ;;    ("HOLD" :foreground "magenta" :weight bold)
        ;;    ("CANCELLED" :foreground "forest green" :weight bold)
        ;;    ("MEETING" :foreground "forest green" :weight bold)
        ;;    ("PHONE" :foreground "forest green" :weight bold))))

(setq org-use-fast-todo-selection t)

(setq org-treat-S-cursor-todo-selection-as-state-change nil)

(after! org
  (setq org-directory "~/org"
        org-default-notes-file "~/org/inbox.org"
        +org-capture-todo-file "~/org/inbox.org"
        +org-capture-journal-file "~/org/journal.org"))

;; I use C-c c to start capture mode
(global-set-key (kbd "C-c c") 'org-capture)

(after! org
  (setq org-capture-templates
        '(("t" "Todo" entry (file +org-capture-todo-file)
           "* TODO %?\n:PROPERTIES:\n:CREATED: %U\n:END:\n")
          ("r" "Responder Email" entry (file +org-capture-todo-file)
               "* NEXT Respond %:from on %:subject\nSCHEDULED: %t\n:PROPERTIES:\n:CREATED: %U\n:END:\n%a\n" :immediate-finish t)
          ("n" "Note" entry (file org-default-notes-file)
               "* %? :NOTE:\n:PROPERTIES:\n:CREATED: %U\n:END:\n%a\n" :clock-in t :clock-resume t)
          ("j" "Journal" entry (file+olp-datetree +org-capture-journal-file)
               "* %U %?\n" :clock-in t :clock-resume t)
          ("w" "org-protocol" entry (file +org-capture-todo-file)
               "* TODO Review %c\n:CREATED: %U\n:END:\n" :immediate-finish t)
          ("m" "Meeting" entry (file +org-capture-notes-file)
               "* MEETING with %?\n:CREATED: %U\n:END:\n" :clock-in t :clock-resume t)
          ("p" "Phone call" entry (file +org-capture-notes-file)
               "* PHONE call with %?\n:CREATED: %U\n:END:\n" :clock-in t :clock-resume t))))

; Targets include this file and any file contributing to the agenda - up to 9 levels deep
(after! org
  (setq org-refile-targets '((nil :maxlevel . 9) (org-agenda-files :maxlevel . 9))
        ; Use full outline paths for refile targets - we file directly with IDO
        org-refile-use-outline-path t
        ; Targets complete directly with IDO
        org-outline-path-complete-in-steps nil
        ; Allow refile to create parent tasks with confirmation
        org-refile-allow-creating-parent-nodes (quote confirm)))

; Exclude DONE state tasks from refile targets
(defun sr/verify-refile-target ()
  "Exclude todo keywords with a done state from refile targets"
  (not (member (nth 2 (org-heading-components)) org-done-keywords)))

(after! org
  (setq org-refile-target-verify-function 'sr/verify-refile-target))

;; Do not dim blocked tasks
(setq org-agenda-dim-blocked-tasks nil)

;; Compact the block agenda view
(setq org-agenda-compact-blocks t)

;; Custom agenda command definitions
(setq org-agenda-custom-commands
      '(("N" "Notes" tags "NOTE"
         ((org-agenda-overriding-header "Notes")
          (org-tags-match-list-sublevels t)))
        ("h" "Habits" tags-todo "STYLE=\"habit\""
         ((org-agenda-overriding-header "Habits")
          (org-agenda-sorting-strategy
           '(todo-state-down effort-up category-keep))))
        (" " "Agenda"
         ((agenda "" nil)
          (tags "REFILE"
                ((org-agenda-overriding-header "Tasks to Refile")
                 (org-tags-match-list-sublevels nil))))
         nil)))

(after! org
  (require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-ellipsis " ▼ "
        org-log-done 'time
        org-hide-emphasis-markers t
        calendar-week-start-day 1
        ))
